### PR TITLE
Tensorrt Mix Precision or INT8 conversion, mix precision almost same size and speed with INT8, but better precision, the converted model have good detection result with mix precision.

### PR DESCRIPTION
**Pull request by ZouJiu1**
1. I have checked for Existing Contributions: Before submitting, my contribution is unique and complementary.

2. No related issues.

new feature 

running model in edge device like nvidia jetson

3. **Purpose**:  I want to running yolov8x.pt in nvidia jetson orix nx 16G device，So I have to convert the trained yolov8x.pt model to tensorrt model. FP16 conversion and inference both are OK, it speed is faster than before, but I want more higher fps. 

So I want to convert it to int8, but when I add codes to ultralytics and convert yolov8x.pt to a int8 engine, it is still ok, but when inference with int8, I find it has no detection result. 

I search so many blogs and github issues, including tensorrt github, I was looking for a reason and a solution, but I can not find one. I try all the official yolov8*.pt with int8, these converted int8 engine have no detection result. But I find a possible solution, It is the mix precision [https://github.com/NVIDIA/TensorRT/blob/main/samples/python/efficientdet/build_engine.py#L188-L218](https://github.com/NVIDIA/TensorRT/blob/main/samples/python/efficientdet/build_engine.py#L188-L218). Mix precision definition is that some layers are FP16, other layers are INT8.

When I use the mix precision to convert all official yolov8*.pt, **keep the first, second, last convolution layers FP16, other layers INT8**, these converted engines have a good detection result in bus.jpg, and it size just bigger a little 200KB, because just three convolution layer use FP16 and others still use int8. So my codes is right, my conversion and calibration procedure are correct. I think yolov8 can add the mix precision to convert model for a better performance.

i. The procedure of converting **a mix precision** engine：

1. download coco [val_2017.zip](http://images.cocodataset.org/zips/val2017.zip) and unzip to val_2017.

2. set calib_input=./val_2017, cache_file=./calibration.cache, **half=True, int8=True**

3. run the script, it will using the mix precision mode convert the model to engine file and infer with bus.jpg.<br>
before running, I marked the 322 line `raise SyntaxError(string + CLI_HELP_MSG) from e` in file ultralytics\\cfg\\__init__.py.

ii. The procedure of converting a **INT8** engine：all same beside this (half=False, int8=True)

the converted models (INT8 or MIX Precision) can download from here [https://www.alipan.com/s/FdfFoPDGCWH](https://www.alipan.com/s/FdfFoPDGCWH) in tensorrt 10.0.0b6. \*mix is a mix precision engine, \*int8 is a INT8 engine, nod is no detection.

mix precision log
```
TensorRT: input "images" with shape(1, 3, 640, 640) DataType.FLOAT
TensorRT: output "output0" with shape(1, 84, 8400) DataType.FLOAT
Mixed-Precision Layer /model.0/conv/Conv set to HALF STRICT data type
Mixed-Precision Layer /model.1/conv/Conv set to HALF STRICT data type
Mixed-Precision Layer /model.22/dfl/conv/Conv set to HALF STRICT data type
TensorRT: building FP16 engine as yolov8x.engine
TensorRT: building INT8 engine as yolov8x.engine
```

------------------------------------------------------------------------------------------------------------------
these engines using mix precision have a good detection result in bus.jpg, and these engines using INT8 have no result.

### difference to [https://github.com/ultralytics/ultralytics/pull/9840](https://github.com/ultralytics/ultralytics/pull/9840)
I change the input and output to FLOAT, not FP16， FLOAT input and output is better.

### use different Calibrator base class
in document [tensorrt/developer-guide](https://docs.nvidia.com/deeplearning/tensorrt/developer-guide/%23enable_int8_c), we can use differenct Calibrator base class in calibrator.py, IInt8EntropyCalibrator2 is suitable for CNN network, IInt8MinMaxCalibrator is suitable for NLP network, should try different calibrator .

int8 conversion reference
[https://github.com/NVIDIA/TensorRT/blob/main/samples/python/detectron2](https://github.com/NVIDIA/TensorRT/blob/main/samples/python/detectron2)

mix precision conversion reference (much better)<br>
Tips: the firstly layer and last layer should be FP16, others can be int8. just try it.
[https://github.com/NVIDIA/TensorRT/blob/main/samples/python/efficientdet](https://github.com/NVIDIA/TensorRT/blob/main/samples/python/efficientdet)

below codes can be used for the first commit [commits/e837cf8d32be1a8b96dc2be3e7c7356b87a5c138](https://github.com/ultralytics/ultralytics/pull/9969/commits/e837cf8d32be1a8b96dc2be3e7c7356b87a5c138)

``` python []
import os
import gc
import sys
sys.path.append(r'E:\work\codeRepo\deploy\common\zj\ultralytics')
from ultralytics import YOLO # newest version from "git clone and git pull"
os.environ['CUDA_VISIBLE_DEVICES'] = '0'

'''
Platform: Window11

Ultralytics YOLOv8.1.44 🚀 Python-3.9.18 torch-2.2.1+cu118 CUDA:0 (NVIDIA GeForce RTX 4070 Ti, 12282MiB)

onnx 1.16.0 opset 17

TensorRT 10.0.0b6:  
https://developer.nvidia.com/downloads/compute/machine-learning/tensorrt/10.0.0/zip/TensorRT-10.0.0.6.Windows10.win10.cuda-11.8.zip

nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2022 NVIDIA Corporation
Built on Wed_Sep_21_10:41:10_Pacific_Daylight_Time_2022
Cuda compilation tools, release 11.8, V11.8.89
Build cuda_11.8.r11.8/compiler.31833905_0
'''

if __name__ == '__main__':
    file = r'yolov8x.pt'
    model = YOLO(file)  # load a pretrained model (recommended for training)
    calib_input = r'E:\work\codeRepo\deploy\jz\val2017'
    cache_file = r'E:\work\codeRepo\deploy\jz\calibration.cache'
    if os.path.exists(cache_file):
        os.remove(cache_file)
    '''
    https://docs.nvidia.com/deeplearning/tensorrt/developer-guide/#enable_int8_c
    To avoid this issue, calibrate with as large a single batch as possible, 
    and ensure that calibration batches are well randomized and have similar distribution.
    '''
    results0 = model.export(format='engine', simplify=True,
                            half=True,
                            int8=True,
                            calib_batch_size=20,
                            calib_num_images=len(os.listdir(calib_input)),
                            calib_input=calib_input,
                            cache_file=cache_file,
                            device=0)
    del model
    gc.collect()
    model = YOLO(r"E:\work\%s"%(file.replace(".pt", ".engine")))
    result = model.predict(
                           'https://ultralytics.com/images/bus.jpg', 
                           save_dir=r'.//', 
                           save=True)
```

I have read the CLA Document and I sign the CLA

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Introducing Mixed Precision and INT8 Makeover for High-Performance Inference! 🚀

### 📊 Key Changes
- 🌈 Extended support for exporting models with mixed precision (FP16 and INT8) and INT8 only modes, enhancing performance and accuracy.
- 🛠 Added a calibration process for INT8 conversion, requiring a set of images to guide the precision conversion.
- 🏗 Infrastructure for processing images in batches during INT8 calibration, ensuring efficiency and effectiveness.
- 📚 Documentation updates and new files to guide users through INT8 conversion and mixed precision usage.

### 🎯 Purpose & Impact
- **Boosted Performance & Efficiency:** Mixed precision support harnesses the benefits of FP16 and INT8 for faster inference with minimal accuracy impact.
- **Greater Flexibility:** Users can now opt for pure INT8 conversion, enjoying reduced model size and latency ideal for edge devices.
- **Enhanced Accuracy:** Calibration process for INT8 conversion aims at maintaining high accuracy, even with reduced precision.
- **User Guidance:** Updated documentation and examples make it easier to leverage these advanced features.

👩‍💻👨‍💻 Both developers and practitioners can now enjoy a streamlined path to high-performance, efficient AI models capable of running swiftly on a wider range of devices, from cloud to edge!

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
This Pull Request (PR) appears to be a placeholder or a mistake, as it contains no code changes. 🤷

### 📊 Key Changes
- None (The PR diff is empty.)

### 🎯 Purpose & Impact
- Since there are no changes, there's no direct impact or benefit from this PR. It's possible it was opened by accident, or for administrative purposes like a discussion or to highlight an issue without proposing a code change. Users and developers won't see any differences in their experience or the functionality as a result of this PR. 👀

### CODE REVIEW

**Comment by glenn-jocher** [comment]
Thank you for expressing your interest! 😊 If you have any specific questions or need further clarification on anything, feel free to reach out. Happy to help!

**Comment by glenn-jocher** [comment]
@Burhan-Q @ZouJiu1 this is really interesting, thanks for the PR! Can you guys create a table with benchmarks for the various quantization method accuracies and speeds, i.e. FP32, FP16, INT8 and this mixed precision PR with speed and accuracies on i.e. COCO val?

```
yolo val model=yolov8n.engine data=coco.yaml batch=1
```

**Comment by glenn-jocher** [comment]
@ZouJiu1 one problem I see here is that your PR is assuming LetterBox preprocessing for all model types, but all TensorRT inference must support all model types, including YOLO Classify with CenterCrop, so the inference code needs to be agnostic to the pre and post processing, which should be all handled by the built-in dataloaders for preprocessing, and in AutoBackend for any customized post-processing needed. 

Currently AutoBackend is the only location in the repo that contains backend-specific code, so ideally all TRT code would be directly in the class or in functions in autobackend.py that can be used in the class.

**Comment by ZouJiu1** [comment]
> @Burhan-Q @ZouJiu1 this is really interesting, thanks for the PR! Can you guys create a table with benchmarks for the various quantization method accuracies and speeds, i.e. FP32, FP16, INT8 and this mixed precision PR with speed and accuracies on i.e. COCO val?
> 
> ```
> yolo val model=yolov8n.engine data=coco.yaml batch=1
> ```

I don't have a gpu computer for myself, Above tests are done  in a GeForce 4070 computer of the company I served, So I think I don't have a gpu computer to finish those tests, maybe you can do these work.

**Comment by ZouJiu1** [comment]
> @ZouJiu1 one problem I see here is that your PR is assuming LetterBox preprocessing for all model types, but all TensorRT inference must support all model types, including YOLO Classify with CenterCrop, so the inference code needs to be agnostic to the pre and post processing, which should be all handled by the built-in dataloaders for preprocessing, and in AutoBackend for any customized post-processing needed.
> 
> Currently AutoBackend is the only location in the repo that contains backend-specific code, so ideally all TRT code would be directly in the class or in functions in autobackend.py that can be used in the class.

I will try to work out this problem in the next week, maybe like you mentioned, add backend-specific code to autobackend.py. So it look like I don't need the tensorrt_int8 directory, I will try to merge calibrator.py and image_batcher.py.

**Comment by glenn-jocher** [comment]
Thank you for pointing this out! I understand the importance of ensuring the preprocessing steps are compatible across all model types for TensorRT inference. I'm on it and will explore integrating a more flexible approach to handle both LetterBox and CenterCrop preprocessing within the AutoBackend framework effectively. I appreciate your guidance on this issue! 🛠🚀

**Comment by ZouJiu1** [comment]
> Thank you for pointing this out! I understand the importance of ensuring the preprocessing steps are compatible across all model types for TensorRT inference. I'm on it and will explore integrating a more flexible approach to handle both LetterBox and CenterCrop preprocessing within the AutoBackend framework effectively. I appreciate your guidance on this issue! 🛠🚀

I have done of this, after one hour I will commit it to this branch.

**Comment by ZouJiu1** [comment]
> Thank you for pointing this out! I understand the importance of ensuring the preprocessing steps are compatible across all model types for TensorRT inference. I'm on it and will explore integrating a more flexible approach to handle both LetterBox and CenterCrop preprocessing within the AutoBackend framework effectively. I appreciate your guidance on this issue! 🛠🚀

No matter which model you use [classify, detect, segment, pose, obb], the tensorrt int8 or mix precision conversion will use the built_in preprocess function and built_in dataloader, letterbox for detect and centercrop for classify. inference doen't need modify autobackend.py, so I just merge the dataloader and preprocess.

I have tested [classify, detect, segment, pose, obb] these model, these conversion and inference are passed, So I think this may be better.

```Python []
import os
import gc
import sys
sys.path.append(r'/home/ZouJiu1/ultralytics')
# sys.path.append(r'/home/ultralytics')
from ultralytics import YOLO # newest version from "git clone and git pull"
os.environ['CUDA_VISIBLE_DEVICES'] = '0'

if __name__ == '__main__':
    file = r'yolov8n.pt'
    file = r'yolov8n-cls.pt'
    file = r'yolov8n-seg.pt'
    # file = r'yolov8n-pose.pt'
    # file = r'yolov8n-obb.pt'
    # task: [classify, detect, segment, pose, obb]
    model = YOLO(file, task='segment')  # load a pretrained model (recommended for training)
    calib_input = r'/home/val2017'
    '''
    https://docs.nvidia.com/deeplearning/tensorrt/developer-guide/#enable_int8_c
    To avoid this issue, calibrate with as large a single batch as possible, 
    and ensure that calibration batches are well randomized and have similar distribution.
    '''
    model.export(format=r"engine", source = calib_input, batch=1, calib_batch=20, 
                 simplify=True, half=True, int8=True, device=0, 
                 imgsz=640 if model.task!='classify' else 224)
    del model
    gc.collect()
    model = YOLO(r"/home/%s"%(file.replace(".pt", ".engine")))
    result = model.predict('https://ultralytics.com/images/bus.jpg',
                           save=True,
                           imgsz=640 if model.task!='classify' else 224)
```

**Comment by glenn-jocher** [comment]
Great to hear you've completed the integration and testing for the various model types with TensorRT INT8 and mixed precision conversions, @ZouJiu1! 🚀 This update significantly enhances the flexibility and usability across different tasks. The built-in preprocessing and dataloader modifications sound like a comprehensive approach to ensure consistency and ease of use.

I appreciate the update on changing the input/output to float32 and the improved inference results. Also, thanks for making the engine files available for download and streamlining the `model.export` function with fewer arguments. It's incredible how these changes make the conversion process more accessible. 

The example code you've shared looks clean and user-friendly, making it easier for others to follow and apply this approach to their projects. I'll be sure to check out the engines you've shared. Thanks for all your hard work on this!

**Comment by glenn-jocher** [comment]
@ZouJiu1 thanks for the updates here. Can you please make some changes:

- We can re-use the `batch` argument as the calibrator batch size, no need to have a second argument here.
- We should not create any new arguments to Exporter, the current self.model should be used for export.
- We can not predefine exact layers as you have in this PR as this is not robust to future changes to the architecture and does not generalize to other architectures like YOLOv9, RTDETR, etc.

Remember, all of the current export code is very generalized to all models.

**Comment by glenn-jocher** [comment]
@ZouJiu1 also not to mention the current hard-coded layer definitions you've created here do not extrapolate even within the YOLOv8 models to different sizes, i.e. YOLOv8n, YOLOv8s. 

Any solution we consider for merging **must generalize (work correctly) across all model architectures (YOLOv8, YOLOv9), all model tasks (Detect, Segment, etc.) and all model sizes, (YOLOv8n, YOLOv8s, etc.)**

**Comment by ZouJiu1** [comment]
@glenn-jocher 
my fork main branch is after the main [https://github.com/ultralytics/ultralytics](https://github.com/ultralytics/ultralytics), so I abandon my commit to keep up the main.

this is my last pull request [https://github.com/ultralytics/ultralytics/pull/10046](https://github.com/ultralytics/ultralytics/pull/10046)

1. the calibrator batch size need a large number, but the onnx batch is smaller even 1, so I think the calib_batch is necessary.  according to [https://docs.nvidia.com/deeplearning/tensorrt/developer-guide/#enable_int8_c](https://docs.nvidia.com/deeplearning/tensorrt/developer-guide/#enable_int8_c)
    ```
    To avoid this issue, calibrate with as large a single batch as possible, 
    and ensure that calibration batches are well randomized and have similar distribution.
    ```

2. the redundant self.Model is removed, I just use the internal self.model in Export. `config.int8_calibrator = EngineCalibrator(cache_file, self.args, self.model)`

3.  those predefined exact layers are removed also, when using mix precision, the first two convolution and last convolution are FP16, others are INT8. No need to care about which model is used and which pypi package version is used.
```
            first_conv = 0
            last_conv = 0
            for i in range(network.num_layers):
                layer = network.get_layer(i)
                if layer.type == trt.LayerType.CONVOLUTION:
                    if first_conv < 2:
                        first_conv += 1
                        network.get_layer(i).precision = trt.DataType.HALF
                        LOGGER.info("Mixed-Precision Layer {} set to HALF STRICT data type".format(layer.name))
                    last_conv = layer
            last_conv.precision = trt.DataType.HALF
            LOGGER.info("Mixed-Precision Layer {} set to HALF STRICT data type".format(last_conv.name))
            LOGGER.info(f"{prefix} building a Mix Precision with FP16 and INT8 engine as {f}")
```
 the image size need to keep the same when exporting to engine and inference with engine.

**Comment by glenn-jocher** [comment]
Hey @ZouJiu1, thanks for the update on your pull request and for aligning your fork with the main branch! 🚀

Regarding your points:
- The distinction between calibrator batch size and ONNX batch size is well-taken. It's clear that for effective INT8 calibration, a larger batch size is beneficial as per NVIDIA's guidelines. This is a good argument for maintaining a separate `calib_batch` parameter when necessary for specific use cases like TensorRT.
  
- Great to hear that the redundancy with `self.model` has been addressed. Using the internal model directly should keep the code cleaner and more maintainable.

- The approach to handling mixed precision dynamically based on the layer's position (first two convolutions and the last one in FP16) sounds like a robust solution. It should indeed generalize well across different models and architectures without needing to hard-code layer specifics.

Just a note on keeping the image size consistent during export and inference: this is crucial for maintaining performance and accuracy, so good call on emphasizing that.

Thanks for the contributions and for keeping the implementation generalized! 👍
